src/dataflow/src/logging/timely.rs

```diff
@@ -500,3 +535,40 @@ fn create_address_row(id: i64, worker: i64, address: &[usize]) -> Row {
 
     address_row
 }
+
+/// Construct a dataflow to check that each operator has a unique address.
+///
+/// The implementation arranges the `address_validation` stream by operator and expects values to be
+/// address hashes. It determines that each operator has a unique address hash and asserts
+/// otherwise.
+fn validate_addresses<G: Scope>(address_validation: &Stream<G, ((usize, u64), G::Timestamp, isize)>)
+where
+    <G as ScopeParent>::Timestamp: Lattice,
+{
+    // Input format: ((op, address_hash), time, delta)
+    let arranged = address_validation
+        .as_collection()
+        .arrange_named::<RowSpine<_, u64, _, _>>("Arrange Addresses validation");
+    let _: Stream<_, ()> = arranged
+        .stream
+        .unary(Pipeline, "Address validation", |_cap, _info| {
+            |input, _output| {
+                input.for_each(|_time, data| {
+                    for wrapper in data.iter() {
+                        let batch = &wrapper;
+                        let mut cursor = batch.cursor();
+                        while let Some(key) = cursor.get_key(batch) {
+                            let mut count = 0;
+                            while let Some(_val) = cursor.get_val(batch) {
+                                // Count number of distinct values
+                                count += 1;
+                                cursor.step_val(batch);
+                            }
+                            assert_eq!(1, count, "Non-unique addresses for operator {}", key);
+                            cursor.step_key(batch);
```

Comment:
Should we be concerned about the risk of this crashing due to a hash collision? What is the probability of `hash_vec` producing a collision ?

Reply:
The data is structured as `(operator_id, hash_of_address)`, which is keyed in the arrangement on `operator_id`. The `hash_of_address` captures the address to each operator on each worker, which is why we don't count the `diff` field but rather add it as `count += 1`.

If there was a hash collision in `hash_of_address`, the logic would still only count a single `hash_of_address` per `operator_id` and we would not determine that the dataflow was constructed incorrectly. If there are no hash collisions, the logic would correctly detect an incorrectly structured dataflow.

I can't say what the probability on a hash collision for hashing vectors is, but I'd hope it to be rather low.